mhz19b: add support for the MH-Z19B CO2 sensor

douardda opened the pull request.

UncleRus commented on components/mhz19b/mhz19b.c:
```diff
@@ -0,0 +1,357 @@
+    if (result == ESP_OK) {
+        // 16-bit CO2 value in response Bytes 2 and 3
+        *co2 = (dev->buf[2] << 8) | dev->buf[3];
+        dev->last_ts = esp_timer_get_time();
```
Please explicitly include <esp_timer.h>. This works now because the esp_timer.h is included via esp_log.h, but this may change in future versions of ESP-IDF.

UncleRus commented on components/mhz19b/mhz19b.c:
```diff
@@ -0,0 +1,357 @@
+
+    // Argument check
+    if (version_len < 5) {
+        return MHZ19B_RESULT_ARGUMENT_ERROR;
```
Better to use standard error codes. ESP_ERR_INVALID_ARGUMENT is fine in this case.

UncleRus commented on components/mhz19b/mhz19b.c:
```diff
@@ -0,0 +1,357 @@
+    return result;
+}
+
+/*!
```
Please move all doxygen comments to the header. Source files (* .c) are not processed by the documentation generator.

douardda commented:
thanks for the quick feedback, I'll fix these ASAP

UncleRus commented:
Also CI is failed, please check `hal/gpio_types.h`. In older versions of ESP-IDF, which we also support, this file located in a different directory.
In this case better to include `driver/gpio.h` instead.

douardda commented:
I believe I've made the requested changes. Now I need the feedback from the CI...

UncleRus commented on components/mhz19b/mhz19b.c:
```diff
@@ -0,0 +1,254 @@
+#include <string.h>
+#include <esp_idf_lib_helpers.h>
+#include <esp_log.h>
+#include <esp_timer.h>.
```
syntax error

douardda replied:
jeez... sorry

UncleRus commented on components/mhz19b/mhz19b.c:
```diff
@@ -0,0 +1,254 @@
+
+esp_err_t mhz19b_init(mhz19b_dev_t *dev, uart_port_t uart_port, gpio_num_t tx_gpio, gpio_num_t rx_gpio)
+{
+    uart_config_t uart_config = {
```
All arguments passed to the functions need to be validated, something like `CHECK_ARG()` in other drivers.

UncleRus commented on components/mhz19b/mhz19b.c:
```diff
@@ -0,0 +1,254 @@
+        .flow_ctrl = UART_HW_FLOWCTRL_DISABLE,
+        .source_clk = UART_SCLK_APB,
+    };
+    uart_driver_install(uart_port, MHZ19B_SERIAL_BUF_LEN * 2, 0, 0, NULL, 0);
```
`uart_driver_install()` returns esp_err_t. Please check it and return error if any (use `CHECK()` macro as in other drivers). Same goes for the rest of the function calls.

trombik commented:
`git rebase` + `git push -f` is good when you are working on work-in-progress branches. but it should be avoided after asking others to review the changes because `rebase` overwrites commit history, and reviews refer to a specific commit or a line in a commit would be nullified. for example, a reviewer needs to take _additional steps_ to verify the code includes fixes for issues pointed out in the review.

(this is my private policy, not the repository's one)

douardda commented:
> `git rebase` + `git push -f` is good when you are working on work-in-progress branches. but it should be avoided after asking others to review the changes because `rebase` overwrites commit history, and reviews refer to a specific commit or a line in a commit would be nullified. for example, a reviewer needs to take _additional steps_ to verify the code includes fixes for issues pointed out in the review.
> 
> (this is my private policy, not the repository's one)

I don't really understand the gh development model, TBH. I usually like to build a clean git history, with individual commits being atomic, deal with only one feature/fix, as small as possible, etc. Using the development model you describe (if I get this right) means stacking revisions like "fix the previous commit" one on another... which in the end is just noise. And since I'm not so familiar with C development any more (and the esp-idf-lib environment) it takes me some iterations to make it right...

But I get it makes it a bit more difficult for reviewers (I miss the mercurial evolve feature so much with git...)

UncleRus commented:
> Using the development model you describe (if I get this right) means stacking revisions like "fix the previous commit" one on another... which in the end is just noise.

Usually these commits are squashed when merging PR.

>  And since I'm not so familiar with C development any more (and the esp-idf-lib environment) it takes me some iterations to make it right...

It's OK.

BTW, I fixed the build for old ESP-IDF versions and added documentation, so now this PR can be merged.
Thank you!

douardda commented:
> > Using the development model you describe (if I get this right) means stacking revisions like "fix the previous commit" one on another... which in the end is just noise.
> 
> Usually these commits are squashed when merging PR.

which something I really don't like with this development model :-)

> 
> > And since I'm not so familiar with C development any more (and the esp-idf-lib environment) it takes me some iterations to make it right...
> 
> It's OK.
> 
> BTW, I fixed the build for old ESP-IDF versions and added documentation, so now this PR can be merged.

seen that, thanks a lot.

> Thank you!
